libgalois/include/katana/EntityTypeManager.h

```diff
@@ -104,28 +109,47 @@ class EntityTypeManager {
     return new_entity_type_id;
   }
 
-  /// adds a new entity type for the set of atomic types with names \p names
+  /// Adds a new entity type for the set of atomic types with names \p names.
+  /// The returned type is the intersection of the types passed in.
+  ///
   /// \returns the EntityTypeID for the new type
-  EntityTypeID AddNonAtomicEntityType(const std::vector<std::string>& names) {
-    for (auto& name : names) {
-      if (!HasAtomicType(name)) {
-        AddAtomicEntityType(name);
-      }
+  template <typename Container>
+  Result<EntityTypeID> GetOrAddNonAtomicEntityTypeFromStrings(
+      const Container& names) {
```

Comment:
I actually meant to remove this. Ooops. We can have it, but I was actually going to propose not having this utility since it can be written inline as: `KATANA_CHECKED(entity_type_manager->GetOrAddNonAtomicEntityType(KATANA_CHECKED(entity_type_manager->GetOrAddEntityTypeIDs(field_names))))`. Is this too ugly for you? (I know the error propagation is ugly, but it's what we get for rolling our own exceptions.)

Reply:
I think APIs like this "pay for themselves" by making the client code easier to read.  I find the alternative....obscure.

Reply:
I don't see why it's obscure you are getting the type IDs with `GetOrAddEntityTypeIDs` and then passing the list to `GetOrAddNonAtomicEntityType`. Seems pretty clear you are getting a non-atomic type for the set of named types you passed in. This is how function composition works.

I can leave this function though. It just seems like a case of multiple ways to do the same thing.

Reply:
I second Emmet in saying that this is required. I'd prefer polymorphism to "FromStrings".

Reply:
Polymorphism isn't possible here since this is a template and the other is not. We could make them all templates, but that would make it a pain to wrap this for Python, when and if we need to.

Reply:
Ok. But we do really need an arbitrary container type?

Reply:
Yea. We seem to. There was code doing conversions between set and vector to pass to this function in older versions of the code. This avoids the conversions and cleans up code.

I guess we could go through and convert everything to set or vector, but I have a hunch that wouldn't work well since the code probably picked one of the other for a reason.